// pigments/lib/utils.js
(function() {
  var utils;

  utils = {
    strip: function(str) {
      return str.replace(/\s+/g, '');
    },
    clamp: function(n) {
      return Math.min(1, Math.max(0, n));
    },
    clampInt: function(n, max) {
      if (max == null) {
        max = 100;
      }
      return Math.min(max, Math.max(0, n));
    },
    parseIntOrPercent: function(value) {
      if (value.indexOf('%') !== -1) {
        return value = Math.round(parseFloat(value) * 2.55);
      } else {
        return value = parseInt(value);
      }
    },
    parseFloatOrPercent: function(amount) {
      if (amount.indexOf('%') !== -1) {
        return parseFloat(amount) / 100;
      } else {
        return parseFloat(amount);
      }
    },
    findClosingIndex: function(s, startIndex, openingChar, closingChar) {
      var curStr, index, nests;
      if (startIndex == null) {
        startIndex = 0;
      }
      if (openingChar == null) {
        openingChar = "[";
      }
      if (closingChar == null) {
        closingChar = "]";
      }
      index = startIndex;
      nests = 1;
      while (nests && index < s.length) {
        curStr = s.substr(index++, 1);
        if (curStr === closingChar) {
          nests--;
        } else if (curStr === openingChar) {
          nests++;
        }
      }
      if (nests === 0) {
        return index - 1;
      } else {
        return -1;
      }
    },
    split: function(s, sep) {
      var a, c, i, l, start;
      if (sep == null) {
        sep = ",";
      }
      a = [];
      l = s.length;
      i = 0;
      start = 0;
      while (i < l) {
        c = s.substr(i, 1);
        switch (c) {
          case "(":
            i = utils.findClosingIndex(s, i + 1, c, ")");
            break;
          case "[":
            i = utils.findClosingIndex(s, i + 1, c, "]");
            break;
          case "":
            i = utils.findClosingIndex(s, i + 1, c, "");
            break;
          case sep:
            a.push(utils.strip(s.substr(start, i - start)));
            start = i + 1;
        }
        i++;
      }
      a.push(utils.strip(s.substr(start, i - start)));
      return a;
    }
  };

  module.exports = utils;

}).call(this);
